theme-utils.mjs

```diff
@@ -1349,6 +1361,102 @@ async function escapePatterns() {
 	}
 }
 
+async function validateThemes( themes ) {
+	function readJson( file ) {
+		return fs.promises.readFile( file, 'utf-8' ).then( JSON.parse );
+	}
+	async function loadSchema( uri ) {
+		if ( ! uri || ! URL.canParse( uri ) ) {
+			return {
+				$schema: 'http://json-schema.org/draft-07/schema#',
+				type: 'object',
+				required: [ '$schema' ],
+			};
+		}
+		const url = new URL( uri );
+		if ( url.protocol === 'http:' || url.protocol === 'https:' ) {
+			return fetch( url ).then( ( res ) => res.json() );
+		}
+		throw new Error( `Unsupported schema protocol: ${ url.protocol }` );
+	}
+	const ajvOptions = {
+		strict: false, // TODO: Evaluate cases where this can be true.
+		allowMatchingProperties: true,
+		allErrors: true,
+		loadSchema,
+	};
+	const ajv = {
+		'http://json-schema.org/draft-07/schema#': new Ajv( ajvOptions ),
+		'http://json-schema.org/draft-04/schema#': new AjvDraft04( ajvOptions ),
+	};
+	const errors = [];
+	const warnings = [];
+	const progress = startProgress( themes.length );
+	for ( const themeSlug of themes ) {
+		const styleCss = await fs.promises.readFile(
+			`${ themeSlug }/style.css`,
+			'utf-8'
+		);
+		const requiresAtLeast = getThemeMetadata(
+			styleCss,
+			'Requires at least'
+		);
+		for ( const file of [
+			`${ themeSlug }/theme.json`,
+			...glob.sync( `${ themeSlug }/styles/*.json` ),
+		] ) {
+			let schemaUri;
+			try {
+				const data = await readJson( file );
+				if ( requiresAtLeast ) {
+					schemaUri = `https://schemas.wp.org/wp/${ requiresAtLeast }/theme.json`;
+					if ( data.$schema !== schemaUri ) {
+						warnings.push( {
+							file,
+							warning: [
+								{
+									actual: data.$schema,
+									expected: schemaUri,
+									validatingWith: schemaUri,
+									message: 'mismatched $schema with "Requires at least" metadata in style.css',
+								},
+							],
+						} );
+					}
+				} else {
+					schemaUri = data.$schema;
+					warnings.push( {
+						file,
+						warning: [
+							{
+								validatingWith: schemaUri,
+								message:
+									'missing "Requires at least" metadata in style.css',
+							},
+						],
+					} );
+				}
+				const schema = await loadSchema( schemaUri );
+				const validate =
+					await ajv[ schema.$schema ].compileAsync( schema );
+				if ( ! validate( data ) ) {
+					throw validate.errors;
+				}
+			} catch ( error ) {
+				errors.push( { file, schema: schemaUri, error } );
+			}
+		}
+		progress.increment();
+	}
+	if ( warnings.length ) {
+		console.dir( warnings, { depth: null } );
+	}
+	if ( errors.length ) {
+		console.dir( errors, { depth: null } );
+		process.exit( 1 );
+	}
```

Comment:
I tried formatting the output into a table. Following is the sample.

```suggestion
	const tableData = []
	warnings.forEach( warning => {
		const {actual, expected, validatingWith, message} = warning.warning[0];
		const padding = 16;
		const file = warning.file + '\n' + chalk.bgBlack.yellow('warning');
		const result = [
			['actual', actual],
			['expected', expected],
			['validating with', validatingWith],
			['message', message],
		].map( line => `${line[0].padEnd( padding )}: ${line[1]}` ).join( '\n' );
		tableData.push([file, result]);
	});

	if ( warnings.length ) {
		// console.dir( warnings, { depth: null } );
		console.log(table(tableData));
	}
```

Reply:
<img width="810" alt="image" src="https://github.com/user-attachments/assets/235a447f-7c3c-4919-8933-05c5249a887e">

Reply:
Fixing this has taken a bit longer than expected. I like the table, but I chose the JSON format to match the errors generated by AJV which are an entire list.

Thinking something like this for the final if I can figure out the rowspan.

```
╔═══════════╤═══════════════════════════════════════════════╗
║ warning   │ ${warning[0][0].key} : ${warning[0][0].value} ║
║ ${file}   │ ${warning[0][1].key} : ${warning[0][1].value} ║
║ ${schema} │ ${warning[0][2].key} : ${warning[0][2].value} ║
╟───────────┼───────────────────────────────────────────────╢
║ error     │ ${error[0][0].key} : ${error[0][0].value}     ║
║ ${file}   │ ${error[0][1].key} : ${error[0][1].value}     ║
║ ${schema} │ ${error[0][2].key} : ${error[0][2].value}     ║
║           ├───────────────────────────────────────────────╢
║           │ ${error[1][0].key} : ${error[1][0].value}     ║
║           │ ${error[1][1].key} : ${error[1][1].value}     ║
║           │ ${error[1][2].key} : ${error[1][2].value}     ║
║           ├───────────────────────────────────────────────╢
║           │ ${error[2][0].key} : ${error[2][0].value}     ║
║           │ ${error[2][1].key} : ${error[2][1].value}     ║
║           │ ${error[2][2].key} : ${error[2][2].value}     ║
╚═══════════╧═══════════════════════════════════════════════╝
```

Reply:
Thanks for accommodating the suggestion. Information is well organized. 💯

Reply:
I may have had a bit too much fun with building the tables and adding colors, so the code exploded a bit. 😅

But I think the new result is _really_ nice. ✨ 

<img width="828" alt="image" src="https://github.com/user-attachments/assets/987d6c2f-b993-4c16-b678-1757752ea820">

And the added `--format=json` option is also super helpful when combined with `jq` to drill down into the data.

This is the breakdown of all the current themes and a count of the types of problems that they have.

```shell-session
$ node theme-utils.mjs validate-theme --format=json $(find . -name 'theme.json' | awk -F/ '{print $2}' | uniq | sort | paste -s -d, -) | jq '.[].data[].message' | sort | uniq -c | sort -bgr
5815 "must NOT have additional properties"
 696 "must be object"
 551 "must be string"
 328 "the $schema version does not match style.css 'Requires at least' version"
 153 "must match exactly one schema in oneOf"
  73 "must be equal to constant"
  71 "the 'Requires at least' version does not support theme.json"
  71 "must be equal to one of the allowed values"
  47 "Missing $schema URI: undefined"
   6 "property name must be valid"
   5 "must be number"
   2 "must be >= 1"
   2 "missing 'Requires at least' header metadata"
   1 "must have required property 'version'"
   1 "must have required property 'slug'"
   1 "must have required property 'name'"
```

Thanks for the push to improve it! 🥇